### ISSUE TITLE
clang-tidy does not complain about missing commas in config file

### ISSUE THREAD

**Issue by tpudlik** [llvm/llvm-project#53737]
The following `.clang-tidy` file will enable certain checks:

```
---
Checks: >
    bugprone-argument-comment,
    bugprone-assert-side-effect,
    bugprone-bool-pointer-implicit-conversion,
...
```

The following `.clang-tidy` file (missing the trailing commas) will enable _no_ checks, but clang-tidy will not complain:

```
---
Checks: >
    bugprone-argument-comment
    bugprone-assert-side-effect
    bugprone-bool-pointer-implicit-conversion
...
```

It is easy to forget to add a trailing comma in a long list, accidentally disabling checks ([real-world example](https://pigweed-review.googlesource.com/c/pigweed/pigweed/+/83565)).  Clang-tidy should perform some validation of the config to catch such bugs.

**Comment by SimplyDanny** [comment]
I've created [a fix](https://reviews.llvm.org/D120187) allowing newline characters as separators of checks.

**Comment by SimplyDanny** [comment]
clang-tidy now considers checks as being distinct if they are either separated by a comma or by a newline character. That means that
```
Checks: >
    bugprone-argument-comment,
    bugprone-assert-side-effect,
    bugprone-bool-pointer-implicit-conversion
```
means the same as
```
Checks: |
    bugprone-argument-comment
    bugprone-assert-side-effect
    bugprone-bool-pointer-implicit-conversion
```
or `Checks: bugprone-argument-comment,bugprone-assert-side-effect,bugprone-bool-pointer-implicit-conversion`. 

clang-tidy will not warn about missing commas. Instead, it will accept such lists more naturally even without commas as separators. Is this an acceptable solution for you, @tpudlik?

**Comment by tpudlik** [comment]
Thank you! Just to clarify, what will happen if someone uses the folded style indicator (`>`), but _forgets_ some of the commas? Example:

```yaml
Checks: >
    bugprone-argument-comment,
    bugprone-assert-side-effect # Note, no comma!
    bugprone-bool-pointer-implicit-conversion,
    yet-another-check
```

The [issue that originally prompted this bug](https://pigweed-review.googlesource.com/c/pigweed/pigweed/+/83565) was of this form (some commas were missing in a long list of checks).

Looking at the test in https://reviews.llvm.org/D120187, it appears this config will now be parsed correctly, but I wanted to double-check.

**Comment by SimplyDanny** [comment]
This specific example will only be correctly parsed if the pipe character `|` is used. The point here is that the YAML parser does its thing and returns simple strings where the `>` and `|` are already resolved. Only then the checks are split at newlines (if `|` was used) or commas or both.

I guess, the general advice now is to go with `|` instead of `>` to avoid such errors.

**Comment by bilderbuchi** [comment]
> clang-tidy now considers checks as being distinct if they are either separated by a comma or by a newline character. That means that
> 
> ```
> Checks: >
>     bugprone-argument-comment,
>     bugprone-assert-side-effect,
>     bugprone-bool-pointer-implicit-conversion
> ```
> 
> means the same as
> 
> ```
> Checks: |
>     bugprone-argument-comment
>     bugprone-assert-side-effect
>     bugprone-bool-pointer-implicit-conversion
> ```
> 
> or `Checks: bugprone-argument-comment,bugprone-assert-side-effect,bugprone-bool-pointer-implicit-conversion`.
> 
> clang-tidy will not warn about missing commas. Instead, it will accept such lists more naturally even without commas as separators.

@SimplyDanny has this fix made it into the codebase correctly?

I'm on 15.0.7, and with a `.clang-tidy` containing
```yaml
---
Checks: |
    bugprone-argument-comment
    bugprone-assert-side-effect
    bugprone-bool-pointer-implicit-conversion
...
```
gives
> clang-tidy  --verify-config
/somepath/.clang-tidy: warning: unknown check 'bugprone-argument-comment
bugprone-assert-side-effect
bugprone-bool-pointer-implicit-conversion' [-verify-config]

while 
```yaml
---
Checks: >
    bugprone-argument-comment,
    bugprone-assert-side-effect,
    bugprone-bool-pointer-implicit-conversion
...
```
gives 
> clang-tidy  --verify-config
No config errors detected.

```
 clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 15.0.7
  Optimized build.
  Default target: x86_64-alpine-linux-musl
  Host CPU: tigerlake
```

**Comment by SimplyDanny** [comment]
The fix/feature made it into the release back than, but as I've just checked it doesn't work anymore.

**Comment by bilderbuchi** [comment]
The test for this [is still in the codebase](https://github.com/llvm/llvm-project/blob/f49b6afc231242dfee027d5da69734836097cd43/clang-tools-extra/unittests/clang-tidy/ClangTidyOptionsTest.cpp#L97-L109). I could not figure out if it is actually still _run_ -- I'm unfamiliar with the LLVM infra, so can't help with that, sorry.

**Comment by SimplyDanny** [comment]
I may have a look. The test is still there and running, but seems to be testing something irrelevant and didn't protect agains breaking the feature.
